src/Compilers/CSharp/Portable/Symbols/PublicModel/NamedTypeSymbol.cs

```diff
@@ -0,0 +1,202 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols.PublicModel
+{
+    internal abstract class NamedTypeSymbol : TypeSymbol, INamedTypeSymbol
+    {
+        private ImmutableArray<ITypeSymbol> _lazyTypeArguments;
+
+        public NamedTypeSymbol(CodeAnalysis.NullableAnnotation nullableAnnotation = CodeAnalysis.NullableAnnotation.None)
+            : base(nullableAnnotation)
+        {
+        }
+
+        internal abstract Symbols.NamedTypeSymbol UnderlyingNamedTypeSymbol { get; }
+
+        int INamedTypeSymbol.Arity
+        {
+            get
+            {
+                return UnderlyingNamedTypeSymbol.Arity;
+            }
+        }
+
+        ImmutableArray<IMethodSymbol> INamedTypeSymbol.InstanceConstructors
+        {
+            get
+            {
+                return UnderlyingNamedTypeSymbol.InstanceConstructors.GetPublicSymbols();
+            }
+        }
+
+        ImmutableArray<IMethodSymbol> INamedTypeSymbol.StaticConstructors
+        {
+            get
+            {
+                return UnderlyingNamedTypeSymbol.StaticConstructors.GetPublicSymbols();
+            }
+        }
+
+        ImmutableArray<IMethodSymbol> INamedTypeSymbol.Constructors
+        {
+            get
+            {
+                return UnderlyingNamedTypeSymbol.Constructors.GetPublicSymbols();
+            }
+        }
+
+        IEnumerable<string> INamedTypeSymbol.MemberNames
+        {
+            get
+            {
+                return UnderlyingNamedTypeSymbol.MemberNames;
+            }
+        }
+
+        ImmutableArray<ITypeParameterSymbol> INamedTypeSymbol.TypeParameters
+        {
+            get
+            {
+                return UnderlyingNamedTypeSymbol.TypeParameters.GetPublicSymbols();
+            }
+        }
+
+        ImmutableArray<ITypeSymbol> INamedTypeSymbol.TypeArguments
+        {
+            get
+            {
+                if (_lazyTypeArguments.IsDefault)
+                {
+
+                    ImmutableInterlocked.InterlockedCompareExchange(ref _lazyTypeArguments, UnderlyingNamedTypeSymbol.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics.SelectAsArray(t => t.GetITypeSymbol()), default);
+                }
+
+                return _lazyTypeArguments;
+            }
+        }
+
+        ImmutableArray<CodeAnalysis.NullableAnnotation> INamedTypeSymbol.TypeArgumentNullableAnnotations
+        {
+            get
+            {
+                return UnderlyingNamedTypeSymbol.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics.SelectAsArray(a => a.ToPublicAnnotation());
```

Comment:
>a => a.ToPublicAnnotation() [](start = 112, length = 27)

I'm not as worried about this one because we intend to deprecate this API.

Reply:
> I'm not as worried about this one because we intend to deprecate this API.

This is how this API was implemented before.

---
In reply to: [338827249](https://github.com/dotnet/roslyn/pull/39498#discussion_r338827249) [](ancestors = 338827249)